skimage/morphology/_misc_cy.pyx

```diff
@@ -0,0 +1,122 @@
+#cython: wraparound=False
+
+"""Cython code used in `remove_objects_by_distance` function."""
+
+cimport numpy as cnp
+
+from .._shared.fused_numerics cimport np_anyint
+
+
+def _remove_objects_by_distance(
+    np_anyint[::1] out not None,
+    Py_ssize_t[::1] border_indices not None,
+    Py_ssize_t[::1] inner_indices not None,
+    kdtree,
+    cnp.float64_t p_norm,
+    cnp.float64_t min_distance,
+    tuple shape,
+):
+    """Remove objects, in specified order, until remaining are a minimum distance apart.
+
+    Remove labeled objects from an image until the remaining ones are spaced
+    more than a given distance from one another. By default, smaller objects
+    are removed first.
+
+    Parameters
+    ----------
+    out :
+        An array with labels for each object in `image` matching it in shape.
+    border_indices, inner_indices :
+        Indices into `out` for the border of objects (`border_indices`) and
+        the inner part of objects (`inner_indices`). `border_indices`
+        determines the iteration order; objects that are indexed first are
+        preserved. Indices must be sorted such, that indices pointing to the
+        same object are next to each other.
+    kdtree : scipy.spatial.cKDTree
+        A KDTree containing the coordinates of all objects in `image`.
+    min_distance :
+        The minimal allowed distance between objects.
+    p_norm :
+        The Minkowski p-norm used to calculate the distance between objects.
+        Defaults to 2 which corresponds to the Euclidean distance.
+    shape :
+        The shape of the unraveled `image`.
+    """
+    cdef:
+        Py_ssize_t i_indices, j_indices  # Loop variables to index `indices`
+        Py_ssize_t i_out  # Loop variable to index `out`
+        np_anyint object_id, other_id
+        list neighborhood
+        set remembered_ids
+
+    remembered_ids = set()
+    for i_indices in range(border_indices.shape[0]):
+        i_out = border_indices[i_indices]
+        object_id = out[i_out]
+        # Skip if sample is part of a removed object
+        if object_id == 0:
+            continue
+
+        neighborhood = kdtree.query_ball_point(
+            kdtree.data[i_indices, ...],
+            r=min_distance,
+            p=p_norm,
+        )
+        for j_indices in neighborhood:
+            # Check object IDs in neighborhood
+            other_id = out[border_indices[j_indices]]
+            if other_id != 0 and other_id != object_id:
+                # If neighbor ID wasn't already removed or is the current one
+                # remove the boundary and remember the ID
+                _remove_object(out, border_indices, j_indices)
+                remembered_ids.add(other_id)
+
+    # Delete inner parts of remembered objects
+    for j_indices in range(inner_indices.shape[0]):
+        object_id = out[inner_indices[j_indices]]
+        if object_id != 0 and object_id in remembered_ids:
+            _remove_object(out, inner_indices, j_indices)
```

Comment:
Is this fast?

I have an alternate implementation for this section that may be worth exploring:

(1) build a coo matrix -> csr matrix mapping object ids (rows) to raveled indices (columns). This can be done with:

```python
sz = label_image.size
coo = sparse.coo_matrix(
    (np.ones(sz), (label_image.ravel(), np.arange(sz))),
    shape=(np.max(label_image), sz + 1),
)
csr = coo.tocsr()
```

(Bonus: `csr.sum(axis=1)` is fast and gives you the bincount)

(2) Finding all the indices belonging to label i is then O(1) and they are adjacent in memory: `idxs = csr.indices[csr.indptr[i]:csr.indptr[i+1]]`

Reply:
(Indeed I expect you should be able to iterate over removed indices in Python rather than Cython without much performance cost, since typically the size of the indices is much greater than the number of labels to be removed.)

Reply:
Indeed re-reading the code, it seems you are iterating over all indices and checking whether they are in the remembered objects — even if the object is a set and O(1) lookup, this seems less efficient than only iterating over remembered object ids and instantly recalling the relevant indices. I think?

Reply:
Sounds interesting! Let me explore this once I find the time!

Reply:
Hmm, is there a typo in the code you are suggesting? Running 

<details><summary>Details</summary>

```python
import numpy as np
import scipy as sp
label_image = np.array(
    [[8, 0, 0, 0, 0, 0, 0, 0, 0, 9, 9],
     [8, 8, 8, 0, 0, 0, 0, 0, 0, 9, 9],
     [0, 0, 0, 0, 0, 0, 0, 0, 9, 0, 0],
     [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
     [0, 0, 3, 0, 0, 0, 1, 0, 0, 0, 0],
     [2, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
     [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
     [0, 0, 0, 0, 0, 0, 0, 0, 0, 7, 7]]
)
sz = label_image.size
coo = sp.sparse.coo_matrix(
    (np.ones(sz), (label_image.ravel(), np.arange(sz))),
    shape=(np.max(label_image), sz + 1),
)
csr = coo.tocsr()
```
</details> 

gives 
```
ValueError: axis 0 index 9 exceeds matrix dimension 9
```
when calling `coo_matrix`...?

Reply:
yes, sorry:

```
shape=(np.max(label_image) + 1, sz),
```

😅 

btw this is the same idea as #4855

Reply:
Thanks, I managed to get it working with

<details><summary>Details</summary>

```patch
Subject: [PATCH] Use sparse matrix to remove objects
---
Index: skimage/morphology/_misc_cy.pyx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/skimage/morphology/_misc_cy.pyx b/skimage/morphology/_misc_cy.pyx
--- a/skimage/morphology/_misc_cy.pyx	(revision d749dd316d40fd7d1ca71c74d170fd16b2c5143d)
+++ b/skimage/morphology/_misc_cy.pyx	(date 1718287031000)
@@ -18,6 +18,7 @@
     cnp.float64_t p_norm,
     cnp.float64_t min_distance,
     tuple shape,
+    csr,
 ):
     """Remove objects, in specified order, until remaining are a minimum distance apart.
 
@@ -70,15 +71,9 @@
             other_id = out[border_indices[j_indices]]
             if other_id != 0 and other_id != object_id:
                 # If neighbor ID wasn't already removed or is the current one
-                # remove the boundary and remember the ID
-                _remove_object(out, border_indices, j_indices)
-                remembered_ids.add(other_id)
+                # remove the boundary
+                out.base[csr.indices[csr.indptr[other_id]:csr.indptr[other_id + 1]]] = 0
 
-    # Delete inner parts of remembered objects
-    for j_indices in range(inner_indices.shape[0]):
-        object_id = out[inner_indices[j_indices]]
-        if object_id != 0 and object_id in remembered_ids:
-            _remove_object(out, inner_indices, j_indices)
 
 
 cdef inline _remove_object(
Index: skimage/morphology/misc.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/skimage/morphology/misc.py b/skimage/morphology/misc.py
--- a/skimage/morphology/misc.py	(revision d749dd316d40fd7d1ca71c74d170fd16b2c5143d)
+++ b/skimage/morphology/misc.py	(date 1718286735125)
@@ -2,6 +2,7 @@
 
 import numpy as np
 import functools
+import scipy as sp
 from scipy import ndimage as ndi
 from scipy.spatial import cKDTree
 
@@ -410,9 +411,16 @@
     # `priority` can only be indexed by positive object IDs,
     # `border_indices` contains all unique sorted IDs so check the lowest / first
     smallest_id = out_raveled[border_indices[0]]
+    largest_id = out_raveled[border_indices[-1]]
     if smallest_id < 0:
         raise ValueError(f"found object with negative ID {smallest_id!r}")
 
+    coo = sp.sparse.coo_matrix(
+        (np.ones(out.size), (out_raveled, np.arange(out.size))),
+        shape=(largest_id + 1, out.size)
+    )
+    csr = coo.tocsr()
+
     try:
         # Sort by priority second using a stable sort to preserve the contiguous
         # sorting of objects. Because each pixel in an object has the same
@@ -447,6 +455,7 @@
         min_distance=min_distance,
         p_norm=p_norm,
         shape=label_image.shape,
+        csr=csr,
     )
 
     if out_raveled.base is not out:

```

</details> 

However, benchmarking with 

```python
import skimage as ski
import scipy as sp
image = ski.data.hubble_deep_field()
image = ski.color.rgb2gray(image)
objects = image > 0.18
label_image, _ = sp.ndimage.label(objects)
%timeit ski.morphology.remove_objects_by_distance(label_image, 1)
%timeit ski.morphology.remove_objects_by_distance(label_image, 5)
%timeit ski.morphology.remove_objects_by_distance(label_image, 100)
%timeit ski.morphology.remove_objects_by_distance(label_image, 500)
```

Shows that the current Cython-loop based approach is still faster.

- Current Cython loop
  364 ms ± 12.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
  346 ms ± 5.39 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
  198 ms ± 2.32 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
  236 ms ± 2.83 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
- Sparse-matrix approach
  379 ms ± 6.25 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
  364 ms ± 746 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)
  224 ms ± 5.08 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
  263 ms ± 881 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)

:thinking:

Reply:
Might still be worth the switch though, as the logic should be way easier to follow...?

Reply:
> the logic should be way easier to follow...?

Yes, *and*, you're currently doing redundant work, because the csr approach is sorting the indices for you. So you should be able to get rid of the argsort calls. (But it might be a bit tricky to deal with border/inner. But should be doable e.g. with a mask I think.)

Reply:
Okay, since this may take me some time to figure out. Do you mind if we added that optimization in another PR? That way we can get this into the next release which we want to do soon. This PR also has kind of a bad track record with getting stalled again because attention shifts again. :sweat_smile: 

cc @jarrodmillman

Reply:
Sure! Might be worth adding a `# TODO` comment to the code pointing to this discussion, but whatever works for you. 😊